### PR TITLE
bugfix-2838-searchfilter-custom-types

### PR DESCRIPTION
**Pull request by farazive**
Added support for doctrine types in ORM SearchFilter.

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | same as branch 2.4
| Fixed tickets | #2838 
| License       | MIT
| Doc PR        | N/A

### CODE REVIEW

**Comment by dunglas** [comment]
The test failure looks related to this patch. Could you try to fix the code?

**Comment by farazive** [comment]
> The test failure looks related to this patch. Could you try to fix the code?

will do

**Comment by rvmourik** [comment]
Any idea when this is going to be available?

**Comment by farazive** [comment]
Terribly sorry guys, I haven't had the time to come back to this. Can someone pick this up please and fix the tests? Thanks.

**Comment by mleczakm** [comment]
Any ETA or docs fix about this, because it is not documented that custom type id's are not supported in filters?

File: src/Bridge/Doctrine/Orm/Filter/SearchFilter.php [hunk 1]
```diff
@@ -155,22 +150,33 @@ protected function filterProperty(string $property, $value, QueryBuilder $queryB
...
+        $queryBuilder->andWhere($orX);
     }
 
     /**
      * Adds where clause according to the strategy.
      *
      * @throws InvalidArgumentException If strategy does not exist
      */
-    protected function addWhereByStrategy(string $strategy, QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $alias, string $field, $value, bool $caseSensitive)
+    protected function addWhereByStrategy(string $strategy, QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $alias, string $field, $value, bool $caseSensitive, $doctrineType)
```
Comment (teohhanhui) [inline src/Bridge/Doctrine/Orm/Filter/SearchFilter.php]: BC: we cannot add a non-optional argument to a `protected` class method.

File: src/Bridge/Doctrine/Orm/Filter/SearchFilter.php [hunk 2]
```diff
@@ -155,22 +150,33 @@ protected function filterProperty(string $property, $value, QueryBuilder $queryB
...
+                ->setParameter($valueParameter, $values[0], $doctrineType);
         } else {
-            $queryBuilder
-                ->andWhere(sprintf('%s.%s IN (:%s)', $associationAlias, $associationField, $valueParameter))
-                ->setParameter($valueParameter, $values);
+            $this->setOrClause($queryBuilder, $queryNameGenerator, $values, $associationAlias, $associationField, $doctrineType, $caseSensitive);
+        }
+    }
+
+    private function setOrClause(QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, array $values, $alias, $field, $doctrineType, $caseSensitive)
```
Comment (teohhanhui) [inline src/Bridge/Doctrine/Orm/Filter/SearchFilter.php]: Note: This method should be removed.

File: src/Bridge/Doctrine/Orm/Filter/SearchFilter.php [hunk 3]
```diff
@@ -155,22 +150,33 @@ protected function filterProperty(string $property, $value, QueryBuilder $queryB
...
+    {
+        $wrapCase = $this->createWrapCase($caseSensitive);
+        $orX = $queryBuilder->expr()->orX();
+        foreach ($values as $value) {
```
Comment (teohhanhui) [inline src/Bridge/Doctrine/Orm/Filter/SearchFilter.php]: We should avoid dynamic queries like this.

**Comment by bettinz** [comment]
Hello, was it fixed on released brand or isn't fixed yet? Thanks 🙏

**Comment by yakobe** [comment]
It would be cool to get this in. Any help needed?

**Comment by alanpoulain** [comment]
Same as https://github.com/api-platform/core/pull/3774.
Had to be reverted (see https://github.com/api-platform/core/pull/4134).

**Pull request closed without merging**